File: WebsiteCSharp/App_Code/CSession.cs
using Microsoft.VisualBasic;
using SchemaAudit;
using Framework;
using SchemaMembership;
using System;
using Framework;

public class CSession : Framework.CSessionBase
{

    public static bool IsAdmin
    {
        get
        {
            return (IsLoggedIn && CUser.Current().IsInRole("Administrators"));
        }
    }

    public static bool IsLoggedIn
    {
        get
        {
            return CUser.IsLoggedIn;
        }
    }

    public static CUser User
    {
        get
        {
            return CUser.Current();
        }
    }

    public static CException PageMessageEx2
    {
        set
        {
            PageMessage = (value.Message + ("\r\n" + ("\r\n" + value.StackTrace)));
            while (!(value.Inner == null))
            {
                value = value.Inner;
                PageMessage +=  (value.Message + ("\r\n" + ("\r\n" + value.StackTrace)));
            }

        }
    }

    public static Exception PageMessageEx
    {
        set
        {
            Exception ex = value;
            CAudit_Error.Log(ex);
            PageMessage = (ex.Message + ("\r\n" + ("\r\n" + ex.StackTrace)));
            while (!(ex.InnerException == null))
            {
                ex = ex.InnerException;
                PageMessage += (ex.Message + ("\r\n" + ("\r\n" + ex.StackTrace)));
            }

        }
    }

    public static string PageMessage
    {
        get
        {
            return GetStr("PageMessage");
        }
        set
        {
            SetStr("PageMessage", value);
        }
    }

    public static CDataSrc Db(ESource source)
    {
        if ((source == ESource.Local))
        {
            return CDataSrc.Default;
        }

        if ((source == ESource.Prod))
        {
            return new CWebSrcBinary(CSession.Home_ProdUrl);
        }

        if ((source == ESource.Other))
        {
            return new CWebSrcBinary(CSession.Home_OtherUrl);
        }

        return null;
    }

    public static CSchemaInfo SchemaInfo(ESource source)
    {
        if ((source == ESource.Local))
        {
            return SchemaLocal;
        }

        if ((source == ESource.Prod))
        {
            return SchemaProd;
        }

        if ((source == ESource.Other))
        {
            return SchemaOther;
        }

        return null;
    }

    public static CDataSrc DbSrc
    {
        get
        {
            return CSession.Db(SourceId);
        }
    }

    public static CDataSrc DbTar
    {
        get
        {
            return CSession.Db(TargetId);
        }
    }

    public static CSchemaInfo SchemaSrc
    {
        get
        {
            return CSession.SchemaInfo(SourceId);
        }
        set
        {
            if ((SourceId == ESource.Local))
            {
                SchemaLocal = null;
            }
            else
            {
                SchemaProd = null;
            }

        }
    }

    public static CSchemaInfo SchemaTar
    {
        get
        {
            return CSession.SchemaInfo(TargetId);
        }
        set
        {
            if ((TargetId == ESource.Local))
            {
                SchemaLocal = null;
            }
            else
            {
                SchemaProd = null;
            }

        }
    }

    public static ESource SourceId
    {
        get
        {
            return (ESource)GetInt("RefSourceId", (int)ESource.Local);
        }
        set
        {
            SetInt("RefSourceId", (int)value);
        }
    }

    public static ESource TargetId
    {
        get
        {
            return (ESource)GetInt("RefTargetId", (int)ESource.Prod);
        }
        set
        {
            SetInt("RefTargetId", (int)value);
        }
    }

    public static CSchemaInfo SchemaLocal
    {
        get
        {
            CSchemaInfo obj = (CSchemaInfo)GetObj("SchemaLocal");
            if ((obj == null))
            {
                obj = CSession.Db(ESource.Local).SchemaInfo();
                SetObj("SchemaLocal", obj);
            }

            return obj;
        }
        set
        {
            SetObj("SchemaLocal", value);
        }
    }

    public static CSchemaInfo SchemaProd
    {
        get
        {
            CSchemaInfo obj = (CSchemaInfo)GetObj("SchemaProd");
            if ((obj == null))
            {
                obj = CSession.Db(ESource.Prod).SchemaInfo();
                SetObj("SchemaProd", obj);
            }

            return obj;
        }
        set
        {
            SetObj("SchemaProd", value);
        }
    }

    public static CSchemaInfo SchemaOther
    {
        get
        {
            CSchemaInfo obj = (CSchemaInfo)GetObj("SchemaOther");
            if ((obj == null))
            {
                obj = CSession.Db(ESource.Other).SchemaInfo();
                SetObj("SchemaOther", obj);
            }

            return obj;
        }
        set
        {
            SetObj("SchemaOther", value);
        }
    }

    public static int Home_ViewOrEdit
    {
        get
        {
            return GetInt("ViewOrEdit", 0);
        }
        set
        {
            SetInt("ViewOrEdit", value);
        }
    }

    public static string Home_DevDir
    {
        get
        {
            return GetStr("Home_DevDir", CPushUpgradeClient_Config.SELF_FOLDER);
        }
        set
        {
            SetStr("Home_DevDir", value);
        }
    }

    public static string Home_ProdHost
    {
        get
        {
            return GetStr("Home_ProdHost", CPushUpgradeClient_Config.DEFAULT_HOSTNAME);
        }
        set
        {
            SetStr("Home_ProdHost", value);
        }
    }

    public static string Home_ProdUrl
    {
        get
        {
            return GetStr("Home_ProdUrl", CPushUpgradeClient_Config.DefaultProdUrl);
        }
        set
        {
            SetStr("Home_ProdUrl", value);
        }
    }

    public static string Home_OtherUrl
    {
        get
        {
            return GetStr("Home_OtherUrl", "http://");
        }
        set
        {
            SetStr("Home_OtherUrl", value);
        }
    }
    public static bool Home_FastHash
    {
        get
        {
            return GetBool("Home_FastHash", true);
        }
        set
        {
            SetBool("Home_FastHash", value);
        }
    }
    public static string Home_RemoteDir
    {
        get
        {
            return GetStr("Home_RemoteDir", string.Empty);
        }
        set
        {
            SetStr("Home_RemoteDir", value);
        }
    }

    public static string Home_Ignore
    {
        get
        {
            return GetStr("Home_Ignore", CPushUpgradeClient_Config.DEFAULT_IGNORES);
        }
        set
        {
            SetStr("Home_Ignore", value);
        }
    }

    public static string TableName
    {
        get
        {
            return GetStr("TableName", "*");
        }
        set
        {
            SetStr("TableName", value);
        }
    }

    public static int Home_Data_FullScan
    {
        get
        {
            return GetInt("Home_Data_FullScan", 0);
        }
        set
        {
            SetInt("Home_Data_FullScan", value);
        }
    }

    public static CAudit_SearchFilters AuditTrailFilters()
    {
        CAudit_SearchFilters filters = ((CAudit_SearchFilters)(Get("AuditTrailFilters")));
        if ((filters == null))
        {
            filters = new CAudit_SearchFilters();
            Set("AuditTrailFilters", filters);
        }

        return filters;
    }

    public static bool SqlIsSelect
    {
        get
        {
            return GetBool("SqlIsSelect");
        }
        set
        {
            SetBool("SqlIsSelect", value);
        }
    }

    public static bool SqlAllClients
    {
        get
        {
            return GetBool("SqlAllClients");
        }
        set
        {
            SetBool("SqlAllClients", value);
        }
    }

    public static string SqlStatement
    {
        get
        {
            return GetStr("SqlStatement");
        }
        set
        {
            SetStr("SqlStatement", value);
        }
    }

    public static string SqlUseConn
    {
        get
        {
            return GetStr("SqlUseConn");
        }
        set
        {
            SetStr("SqlUseConn", value);
        }
    }

    public static bool SqlUseProd
    {
        get
        {
            return GetBool("SqlUseProd");
        }
        set
        {
            SetBool("SqlUseProd", value);
        }
    }
}

File: Website/App_Code/CSession.vb
Imports Microsoft.VisualBasic
Imports SchemaAudit

Public Class CSession : Inherits Framework.CSessionBase

#Region "Current Login"
    Public Shared ReadOnly Property IsAdmin() As Boolean
        Get
            Return IsLoggedIn AndAlso CUser.Current.IsInRole("Administrators")
        End Get
    End Property
    Public Shared ReadOnly Property IsLoggedIn() As Boolean
        Get
            Return CUser.IsLoggedIn
        End Get
    End Property
    Public Shared ReadOnly Property User() As CUser
        Get
            Return CUser.Current
        End Get
    End Property
#End Region

#Region "PageMessage"
    Public Shared WriteOnly Property PageMessageEx2 As CException
        Set(ex As CException)
            PageMessage = ex.Message & vbCrLf & vbCrLf & ex.StackTrace
            While Not IsNothing(ex.Inner)
                ex = ex.Inner
                PageMessage &= ex.Message & vbCrLf & vbCrLf & ex.StackTrace
            End While
        End Set
    End Property
    Public Shared WriteOnly Property PageMessageEx As Exception
        Set(ex As Exception)
            CAudit_Error.Log(ex)

            PageMessage = ex.Message & vbCrLf & vbCrLf & ex.StackTrace
            While Not IsNothing(ex.InnerException)
                ex = ex.InnerException
                PageMessage &= ex.Message & vbCrLf & vbCrLf & ex.StackTrace
            End While
        End Set
    End Property
    Public Shared Property PageMessage() As String
        Get
            Return GetStr("PageMessage")
        End Get
        Set(ByVal value As String)
            SetStr("PageMessage", value)
        End Set
    End Property
#End Region

#Region "Self"
    Public Shared Function Db(source As ESource) As CDataSrc
        If source = ESource.Local Then Return CDataSrc.Default
        If source = ESource.Prod Then Return New CWebSrcBinary(CSession.Home_ProdUrl)
        If source = ESource.Other Then Return New CWebSrcBinary(CSession.Home_OtherUrl)
        Return Nothing
    End Function
    Public Shared Function SchemaInfo(source As ESource) As CSchemaInfo
        If source = ESource.Local Then Return SchemaLocal
        If source = ESource.Prod Then Return SchemaProd
        If source = ESource.Other Then Return SchemaOther
        Return Nothing
    End Function

    Public Shared ReadOnly Property DbSrc As CDataSrc
        Get
            Return Db(SourceId)
        End Get
    End Property
    Public Shared ReadOnly Property DbTar As CDataSrc
        Get
            Return Db(TargetId)
        End Get
    End Property

    Public Shared Property SchemaSrc As CSchemaInfo
        Get
            Return SchemaInfo(SourceId)
        End Get
        Set(value As CSchemaInfo)
            If SourceId = ESource.Local Then
                SchemaLocal = Nothing
            Else
                SchemaProd = Nothing
            End If
        End Set
    End Property
    Public Shared Property SchemaTar As CSchemaInfo
        Get
            Return SchemaInfo(TargetId)
        End Get
        Set(value As CSchemaInfo)
            If TargetId = ESource.Local Then
                SchemaLocal = Nothing
            Else
                SchemaProd = Nothing
            End If
        End Set
    End Property


    Public Shared Property SourceId() As ESource
        Get
            Return GetInt("RefSourceId", ESource.Local)
        End Get
        Set(ByVal value As ESource)
            SetInt("RefSourceId", value)
        End Set
    End Property
    Public Shared Property TargetId() As ESource
        Get
            Return GetInt("RefTargetId", ESource.Prod)
        End Get
        Set(ByVal value As ESource)
            SetInt("RefTargetId", value)
        End Set
    End Property


    Public Shared Property SchemaLocal() As CSchemaInfo
        Get
            Dim obj As CSchemaInfo = GetObj("SchemaLocal")
            If IsNothing(obj) Then
                obj = Db(ESource.Local).SchemaInfo
                SetObj("SchemaLocal", obj)
            End If
            Return obj
        End Get
        Set(ByVal value As CSchemaInfo)
            SetObj("SchemaLocal", value)
        End Set
    End Property
    Public Shared Property SchemaProd() As CSchemaInfo
        Get
            Dim obj As CSchemaInfo = GetObj("SchemaProd")
            If IsNothing(obj) Then
                obj = Db(ESource.Prod).SchemaInfo
                SetObj("SchemaProd", obj)
            End If
            Return obj
        End Get
        Set(ByVal value As CSchemaInfo)
            SetObj("SchemaProd", value)
        End Set
    End Property
    Public Shared Property SchemaOther() As CSchemaInfo
        Get
            Dim obj As CSchemaInfo = GetObj("SchemaOther")
            If IsNothing(obj) Then
                obj = Db(ESource.Other).SchemaInfo
                SetObj("SchemaOther", obj)
            End If
            Return obj
        End Get
        Set(ByVal value As CSchemaInfo)
            SetObj("SchemaOther", value)
        End Set
    End Property



    Public Shared Property Home_ViewOrEdit() As Integer
        Get
            Return GetInt("ViewOrEdit", 0)
        End Get
        Set(ByVal value As Integer)
            SetInt("ViewOrEdit", value)
        End Set
    End Property
    Public Shared Property Home_DevDir() As String
        Get
            Return GetStr("Home_DevDir", CPushUpgradeClient_Config.SELF_FOLDER)
        End Get
        Set(ByVal value As String)
            SetStr("Home_DevDir", value)
        End Set
    End Property
    Public Shared Property Home_ProdHost() As String
        Get
            Return GetStr("Home_ProdHost", CPushUpgradeClient_Config.DEFAULT_HOSTNAME)
        End Get
        Set(ByVal value As String)
            SetStr("Home_ProdHost", value)
        End Set
    End Property
    Public Shared Property Home_RemoteDir() As String
        Get
            Return GetStr("Home_RemoteDir")
        End Get
        Set(ByVal value As String)
            SetStr("Home_RemoteDir", value)
        End Set
    End Property
    Public Shared Property Home_ProdUrl() As String
        Get
            Return GetStr("Home_ProdUrl", CPushUpgradeClient_Config.DefaultProdUrl)
        End Get
        Set(ByVal value As String)
            SetStr("Home_ProdUrl", value)
        End Set
    End Property
    Public Shared Property Home_OtherUrl() As String
        Get
            Return GetStr("Home_OtherUrl", "http://")
        End Get
        Set(ByVal value As String)
            SetStr("Home_OtherUrl", value)
        End Set
    End Property
    Public Shared Property Home_Ignore() As String
        Get
            Return GetStr("Home_Ignore", CPushUpgradeClient_Config.DEFAULT_IGNORES)
        End Get
        Set(ByVal value As String)
            SetStr("Home_Ignore", value)
        End Set
    End Property
    Public Shared Property Home_FastHash() As Boolean
        Get
            Return GetBool("Home_FastHash", False)
        End Get
        Set(ByVal value As Boolean)
            SetBool("Home_FastHash", value)
        End Set
    End Property

    Public Shared Property TableName() As String
        Get
            Return GetStr("TableName", "*")
        End Get
        Set(ByVal value As String)
            SetStr("TableName", value)
        End Set
    End Property
    Public Shared Property Home_Data_FullScan() As Integer
        Get
            Return GetInt("Home_Data_FullScan", 0)
        End Get
        Set(ByVal value As Integer)
            SetStr("Home_Data_FullScan", value)
        End Set
    End Property
#End Region


#Region "Search Filters - AuditTrail"
    Public Shared Function AuditTrailFilters() As CAudit_SearchFilters
        Dim filters As CAudit_SearchFilters = CType([Get]("AuditTrailFilters"), CAudit_SearchFilters)
        If IsNothing(filters) Then
            filters = New CAudit_SearchFilters
            [Set]("AuditTrailFilters", filters)
        End If
        Return filters
    End Function
#End Region

#Region "Sql - Current Query"
    Public Shared Property SqlIsSelect() As Boolean
        Get
            Return GetBool("SqlIsSelect")
        End Get
        Set(ByVal value As Boolean)
            SetBool("SqlIsSelect", value)
        End Set
    End Property
    Public Shared Property SqlAllClients() As Boolean
        Get
            Return GetBool("SqlAllClients")
        End Get
        Set(ByVal value As Boolean)
            SetBool("SqlAllClients", value)
        End Set
    End Property
    Public Shared Property SqlStatement() As String
        Get
            Return GetStr("SqlStatement")
        End Get
        Set(ByVal value As String)
            SetStr("SqlStatement", value)
        End Set
    End Property
    Public Shared Property SqlUseConn() As String
        Get
            Return GetStr("SqlUseConn")
        End Get
        Set(ByVal value As String)
            SetStr("SqlUseConn", value)
        End Set
    End Property

    Public Shared Property SqlUseProd() As Boolean
        Get
            Return GetBool("SqlUseProd")
        End Get
        Set(ByVal value As Boolean)
            SetBool("SqlUseProd", value)
        End Set
    End Property

#End Region

End Class
